# Recipe Recommendation System 
~ TASTY DISHES ~

- Group 3
- Group Members.
    - Cindy Tumaini
    - Margret Namunyak
    - Faith Wafula
    - Martin Waweru
    - Matthew Karani


## Table Of Contents

- Business Understanding
- Data Understanding
- Data Preparation
- Modelling 
- Evaluation 
  

## Business Understanding

### Business Description 
Tasty Dishes is a web-based culinary platform dedicated to sharing authentic African recipes with the world. Our mission is to enhance the cooking experience of home chefs by providing them with a diverse collection of recipes rooted in African culinary traditions, while also incorporating global influences. Whether you're an experienced cook or just starting, Tasty Dishes offers a wide variety of recipes that empower users to create delicious, flavorful meals from the comfort of their homes.


## Business Goal 
### Objective
The main objective of this project is to develop an item-based recipe recommendation system that suggests recipes to users based on the ingredients they have available. By analyzing the ingredients present in various recipes, the system aims to provide relevant and appealing recommendations that encourage users to explore and cook diverse dishes rooted in African culinary traditions, while also incorporating global flavors.

### Scope

1. Ingredient-Based Recommendations: Develop an algorithm that analyzes user-provided ingredients to recommend recipes based on ingredient similarity, leveraging a diverse dataset that includes recipe_Title, Ingredients, and Instructions for authentic African and global dishes.

2. User-Friendly Interface: Design an intuitive web interface that enables users to input their available ingredients and view tailored recipe recommendations, along with detailed cooking instructions and a feedback mechanism to enhance recommendation accuracy.


### Success Criteria
1. Accuracy:
Achieve at least 80% accuracy in recommending relevant recipes based on user-provided ingredients.

2. Precision:
Ensure that at least 75% of recommended recipes correspond to the user’s input ingredients.

3. Recall:
Aim for a recall rate of at least 70%, indicating the system identifies a significant portion of relevant recipes.

4. F1 Score:
Target an F1 score of 0.75 or higher, balancing precision and recall for comprehensive recommendations.






## Data Understanding

### Data Source:




In [43]:
# Necessary Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import string
import re
import nltk
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


### Data Frame One

In [44]:
# Load the dataframe

df = pd.read_csv("Food Ingredients and Recipe Dataset with Image Name Mapping.csv", index_col=0)


# Display the first columns
display(df.head(10))

#show the shape
print(df.shape)

Title  \
0  Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                    Crispy Salt and Pepper Potatoes   
2                        Thanksgiving Mac and Cheese   
3                 Italian Sausage and Bread Stuffing   
4                                       Newton's Law   
5                                       Warm Comfort   
6                                 Apples and Oranges   
7                                 Turmeric Hot Toddy   
8                            Instant Pot Lamb Haleem   
9     Spiced Lentil and Caramelized Onion Baked Eggs   

                                         Ingredients  \
0  ['1 (3½–4-lb.) whole chicken', '2¾ tsp. kosher...   
1  ['2 large egg whites', '1 pound new potatoes (...   
2  ['1 cup evaporated milk', '1 cup whole milk', ...   
3  ['1 (¾- to 1-pound) round Italian loaf, cut in...   
4  ['1 teaspoon dark brown sugar', '1 teaspoon ho...   
5  ['2 chamomile tea bags', '1½ oz. reposado tequ...   
6  ['3 oz. Grand Marnier', '1 oz. Amaro Averna', ...   
7  ['¼ cup granulated sugar', '¾ tsp. ground turm...   
8  ['¾ cup assorted dals (such as chana dal, moon...   
9  ['1 (14.5-ounce) can basic lentil soup, like A...   

                                        Instructions  \
0  Pat chicken dry with paper towels, season all ...   
1  Preheat oven to 400°F and line a rimmed baking...   
2  Place a rack in middle of oven; preheat to 400...   
3  Preheat oven to 350°F with rack in middle. Gen...   
4  Stir together brown sugar and hot water in a c...   
5  Place 2 chamomile tea bags in a heatsafe vesse...   
6  Add 3 oz. Grand Marnier, 1 oz. Amaro Averna, a...   
7  For the turmeric syrup, combine ½ cup hot wate...   
8  Combine dals, rice, and barley in a medium bow...   
9  Place an oven rack in the center of the oven, ...   

                                          Image_Name  \
0  miso-butter-roast-chicken-acorn-squash-panzanella   
1         crispy-salt-and-pepper-potatoes-dan-kluger   
2         thanksgiving-mac-and-cheese-erick-williams   
3          italian-sausage-and-bread-stuffing-240559   
4                 newtons-law-apple-bourbon-cocktail   
5               warm-comfort-tequila-chamomile-toddy   
6                    apples-and-oranges-spiked-cider   
7                  turmeric-hot-toddy-claire-sprouse   
8                            instant-pot-lamb-haleem   
9     spiced-lentil-and-caramelized-onion-baked-eggs   

                                 Cleaned_Ingredients  
0  ['1 (3½–4-lb.) whole chicken', '2¾ tsp. kosher...  
1  ['2 large egg whites', '1 pound new potatoes (...  
2  ['1 cup evaporated milk', '1 cup whole milk', ...  
3  ['1 (¾- to 1-pound) round Italian loaf, cut in...  
4  ['1 teaspoon dark brown sugar', '1 teaspoon ho...  
5  ['2 chamomile tea bags', '1½ oz. reposado tequ...  
6  ['3 oz. Grand Marnier', '1 oz. Amaro Averna', ...  
7  ['¼ cup granulated sugar', '¾ tsp. ground turm...  
8  ['¾ cup assorted dals (such as chana dal, moon...  
9  ['1 (14.5-ounce) can basic lentil soup, like A...

(13501, 5)


- Check for duplicates


In [45]:
print(f'Number of duplicates: {df.duplicated().sum()}')

Number of duplicates: 0


In [46]:
#drop duplicates
df.drop_duplicates(inplace=True)
print(f'Number of duplicates after dropping: {df.duplicated().sum()}')

Number of duplicates after dropping: 0


- Check for missing values

In [47]:
df.isnull().sum().sort_values(ascending=False)

Instructions           8
Title                  5
Ingredients            0
Image_Name             0
Cleaned_Ingredients    0
dtype: int64

In [48]:
#drop rows with missing values
df.dropna(inplace=True)
print(f'Number of missing values after dropping: {df.isnull().sum().sum()}')

Number of missing values after dropping: 0


- There is the Ingredients and Cleaned Ingredients column, check if there is any difference between the two.

In [49]:
df['Ingredients'][5]

"['2 chamomile tea bags', '1½ oz. reposado tequila', '¾ oz. fresh lemon juice', '1 Tbsp. agave nectar']"

In [50]:
df['Cleaned_Ingredients'][5]

"['2 chamomile tea bags', '1½ oz. reposado tequila', '¾ oz. fresh lemon juice', '1 Tbsp. agave nectar']"

- There's no significant difference between Ingredients and cleaned Ingredients. Thus, we drop the Ingredients column and rename cleaned_ingredients  ingredients.

In [51]:
#move the cleaned_ingredients column to the second column
df = df[['Title', 'Cleaned_Ingredients', 'Ingredients', 'Instructions', 'Image_Name']]

#drop the ingredients column
df = df.drop(columns=['Ingredients','Image_Name'])


In [52]:
# rename the cleaned ingredients column
df = df.rename(columns={'Cleaned_Ingredients':'Ingredients'})
df.head()


Title  \
0  Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                    Crispy Salt and Pepper Potatoes   
2                        Thanksgiving Mac and Cheese   
3                 Italian Sausage and Bread Stuffing   
4                                       Newton's Law   

                                         Ingredients  \
0  ['1 (3½–4-lb.) whole chicken', '2¾ tsp. kosher...   
1  ['2 large egg whites', '1 pound new potatoes (...   
2  ['1 cup evaporated milk', '1 cup whole milk', ...   
3  ['1 (¾- to 1-pound) round Italian loaf, cut in...   
4  ['1 teaspoon dark brown sugar', '1 teaspoon ho...   

                                        Instructions  
0  Pat chicken dry with paper towels, season all ...  
1  Preheat oven to 400°F and line a rimmed baking...  
2  Place a rack in middle of oven; preheat to 400...  
3  Preheat oven to 350°F with rack in middle. Gen...  
4  Stir together brown sugar and hot water in a c...

In [53]:
df['Ingredients'][5]

"['2 chamomile tea bags', '1½ oz. reposado tequila', '¾ oz. fresh lemon juice', '1 Tbsp. agave nectar']"

### DataFrame two 

In [54]:
#explore the recipeslmp.csv file
df2 = pd.read_csv("RecipesImp.csv")
display(df2.head())

#display the shape
print(df2.shape)

title  index  page  \
0                           Kaimati(Fried Dumplings)  15003    24   
1                      Mahamri\r\n(Swahili Doughnut)  15004    26   
2  Enriched Mandazi \r\n(East African Doughnuts, ...  15124    28   
3  Basic Mandazi \r\n(East African Doughnuts, Basic)  15125    30   
4                  Meat Samosa\r\n(Sambusa ya Nyama)  15025    32   

                                               about  \
0  Kaimatis get their unique flavour from the sty...   
1  This is a typical traditional recipe among the...   
2  A popular snack among urban dwellers across th...   
3  You will find this recipe in any home across K...   
4  Nothing more delicious like the Kenyan meaty s...   

                                         ingridients  \
0   wheat flour, refined\r\nwater, vanilla essenc...   
1  wheat flour,\r\ncoconut milk\r\nwhite sugar\r\...   
2  self-raising wheat flour\r\neggs\r\nmargarine\...   
3  all-purpose wheat flour\r\nbaking powder\r\nsu...   
4  minced beef\r\ncoriander, fresh\r\nleek\r\ngar...   

                                         preparation  \
0   Put yeast in a small container.\r\n Add 50ml ...   
1    Break the coconut shell, drain the water and...   
2  ?  Put flour, salt, sugar and lemon rind into ...   
3  ?  Put the wheat flour into a bowl, add baking...   
4  ?  Put the meat in a pan over a fire. Stir con...   

                        nutrition per 100g of recipe  energy(kcal)  fat(g)  \
0  Energy 1,795 kJ/ 429 kcal | Fat 21.8 g | Carbo...         429.0    21.8   
1  Energy 1,728 kJ/ 413 kcal | Fat 22.1 g | Carbo...         413.0    22.1   
2  Energy 1,590 kJ/ 379 kcal | Fat 16.1 g | Carbo...         379.0    16.1   
3  Energy 1,430kJ/ 340 kcal | Fat 12.9 g | Carboh...         340.0    12.9   
4  Energy 1,854 kJ/ 443 kcal | Fat 22.2 g | Carbo...         443.0    22.2   

   carbohydrates(g)  proteins(g)  fibre(g) vitamin A(mcg)  iron(mg)  zinc(mg)  \
0              52.8          4.6       1.6             30       2.1      0.45   
1              46.6          6.0       2.1             41       2.8      0.56   
2              49.9          7.6       2.2             90       3.3      0.66   
3              48.7          6.4       2.1             48       3.5      0.52   
4              40.5         18.8       3.1             66      11.5      2.99   

   F_factor_est  
0           0.4  
1           0.4  
2           0.4  
3           0.4  
4           0.4

(142, 16)


- Since we want only a few columns to recommend the possible recipes, we need to drop some columns.

In [55]:
df2.columns


Index(['title', 'index', 'page', 'about', 'ingridients', 'preparation',
       'nutrition per 100g of recipe', 'energy(kcal)', 'fat(g)',
       'carbohydrates(g)', 'proteins(g)', 'fibre(g)', 'vitamin A(mcg)',
       'iron(mg)', 'zinc(mg)', 'F_factor_est'],
      dtype='object')

In [56]:
columns_to_keep = ['title','ingridients','preparation']

df2 = df2[columns_to_keep]
df2.head()

title  \
0                           Kaimati(Fried Dumplings)   
1                      Mahamri\r\n(Swahili Doughnut)   
2  Enriched Mandazi \r\n(East African Doughnuts, ...   
3  Basic Mandazi \r\n(East African Doughnuts, Basic)   
4                  Meat Samosa\r\n(Sambusa ya Nyama)   

                                         ingridients  \
0   wheat flour, refined\r\nwater, vanilla essenc...   
1  wheat flour,\r\ncoconut milk\r\nwhite sugar\r\...   
2  self-raising wheat flour\r\neggs\r\nmargarine\...   
3  all-purpose wheat flour\r\nbaking powder\r\nsu...   
4  minced beef\r\ncoriander, fresh\r\nleek\r\ngar...   

                                         preparation  
0   Put yeast in a small container.\r\n Add 50ml ...  
1    Break the coconut shell, drain the water and...  
2  ?  Put flour, salt, sugar and lemon rind into ...  
3  ?  Put the wheat flour into a bowl, add baking...  
4  ?  Put the meat in a pan over a fire. Stir con...

In [57]:
#clean the column names
#change the ingridient column name to ingredients
df2.rename(columns={'ingridients':'ingredients','preparation':'instructions'}, inplace=True)

#capitalize the column names
df2.columns = df2.columns.str.capitalize()

# Function to process the Ingredients column
def process_ingredients(ingredients):
    # Remove unwanted characters and split by commas
    return [ingredient.strip() for ingredient in ingredients.replace('[','').replace(']','').replace("'", "").replace('\n', ',').split(',')]

# Apply the function to each DataFrame
df['Ingredients'] = df['Ingredients'].apply(lambda x: [ingredient.strip() for ingredient in x.split(',')])
df2['Ingredients'] = df2['Ingredients'].apply(process_ingredients)




In [58]:
# Check the processed DataFrames
df[['Title', 'Ingredients']]


Title  \
0      Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                        Crispy Salt and Pepper Potatoes   
2                            Thanksgiving Mac and Cheese   
3                     Italian Sausage and Bread Stuffing   
4                                           Newton's Law   
...                                                  ...   
13496                               Brownie Pudding Cake   
13497  Israeli Couscous with Roasted Butternut Squash...   
13498  Rice with Soy-Glazed Bonito Flakes and Sesame ...   
13499                                        Spanakopita   
13500  Mexican Poblano, Spinach, and Black Bean "Lasa...   

                                             Ingredients  
0      [['1 (3½–4-lb.) whole chicken', '2¾ tsp. koshe...  
1      [['2 large egg whites', '1 pound new potatoes ...  
2      [['1 cup evaporated milk', '1 cup whole milk',...  
3      [['1 (¾- to 1-pound) round Italian loaf, cut i...  
4      [['1 teaspoon dark brown sugar', '1 teaspoon h...  
...                                                  ...  
13496  [['1 cup all-purpose flour', '2/3 cup unsweete...  
13497  [['1 preserved lemon', '1 1/2 pound butternut ...  
13498  [['Leftover katsuo bushi (dried bonito flakes)...  
13499  [['1 stick (1/2 cup) plus 1 tablespoon unsalte...  
13500  [['12 medium to large fresh poblano chiles (2 ...  

[13493 rows x 2 columns]

In [59]:
df2[['Title', 'Ingredients']]

Title  \
0                             Kaimati(Fried Dumplings)   
1                        Mahamri\r\n(Swahili Doughnut)   
2    Enriched Mandazi \r\n(East African Doughnuts, ...   
3    Basic Mandazi \r\n(East African Doughnuts, Basic)   
4                    Meat Samosa\r\n(Sambusa ya Nyama)   
..                                                 ...   
137                   Bhature\r\n (Fried Indian Bread)   
138  Vimumunya vya \r\nSukari\r\n (Sweetened Pumpki...   
139                         Siro\r\n (Semolina & Nuts)   
140                         Chaas\r\n(Diluted Yoghurt)   
141                                    Groundnut Sauce   

                                           Ingredients  
0    [wheat flour, refined, water, vanilla essence,...  
1    [wheat flour, , coconut milk, white sugar, dry...  
2    [self-raising wheat flour, eggs, margarine, Ri...  
3    [all-purpose wheat flour, baking powder, sugar...  
4    [minced beef, coriander, fresh, leek, garlic, ...  
..                                                 ...  
137  [wheat flour, salt, sugar, ghee, cooking oil, ...  
138    [pumpkin, cardamon, sugar, coconut milk, water]  
139  [semolina flour, cow ghee, cow milk, sugar, pi...  
140                   [natural yoghurt, water, salt, ]  
141                [groundnut, salt, sour milk, water]  

[142 rows x 2 columns]

- Chek for missing values

In [60]:
print(f'number of missing values: {df2.isnull().sum().sum()}')

number of missing values: 0


- Since both dataframes have no missing values and duplicates we can merge them now

In [61]:
#merge the two dataframes
combined_df = pd.concat([df,df2])
#check the shapes of the three dfs
print(f'Dataframe 1 has a shape of: {df.shape}')
print(f'Dataframe 2 has a shape of: {df2.shape}')
print(f'Combined dataframe has a shape of: {combined_df.shape}')

#reset the index
combined_df = combined_df.reset_index(drop=True)



Dataframe 1 has a shape of: (13493, 3)
Dataframe 2 has a shape of: (142, 3)
Combined dataframe has a shape of: (13635, 3)


### Clean the Combined DataFrame

In [62]:
combined_df.head()

Title  \
0  Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                    Crispy Salt and Pepper Potatoes   
2                        Thanksgiving Mac and Cheese   
3                 Italian Sausage and Bread Stuffing   
4                                       Newton's Law   

                                         Ingredients  \
0  [['1 (3½–4-lb.) whole chicken', '2¾ tsp. koshe...   
1  [['2 large egg whites', '1 pound new potatoes ...   
2  [['1 cup evaporated milk', '1 cup whole milk',...   
3  [['1 (¾- to 1-pound) round Italian loaf, cut i...   
4  [['1 teaspoon dark brown sugar', '1 teaspoon h...   

                                        Instructions  
0  Pat chicken dry with paper towels, season all ...  
1  Preheat oven to 400°F and line a rimmed baking...  
2  Place a rack in middle of oven; preheat to 400...  
3  Preheat oven to 350°F with rack in middle. Gen...  
4  Stir together brown sugar and hot water in a c...

In [63]:
combined_df[-10:]

Title  \
13625           Vinolo\r\n(Banana and Maize Flour Ugali)   
13626                      Finger Millet \r\nFlour Ugali   
13627                                      White Chapati   
13628                                      Brown Chapati   
13629                          Roti \r\n(Indian Chapati)   
13630                   Bhature\r\n (Fried Indian Bread)   
13631  Vimumunya vya \r\nSukari\r\n (Sweetened Pumpki...   
13632                         Siro\r\n (Semolina & Nuts)   
13633                         Chaas\r\n(Diluted Yoghurt)   
13634                                    Groundnut Sauce   

                                             Ingredients  \
13625                 [banana green, maize flour, water]   
13626                             [finger millet, water]   
13627     [wheat flour, water, sugar, salt, cooking oil]   
13628   [wheat flour, water, sugar, , salt, cooking oil]   
13629  [wheat flour, salt, water, cooking oil, cow ghee]   
13630  [wheat flour, salt, sugar, ghee, cooking oil, ...   
13631    [pumpkin, cardamon, sugar, coconut milk, water]   
13632  [semolina flour, cow ghee, cow milk, sugar, pi...   
13633                   [natural yoghurt, water, salt, ]   
13634                [groundnut, salt, sour milk, water]   

                                            Instructions  
13625  Preparation 5 minutes | Cooking 40 minutes | \...  
13626  Preparation time 5 minutes | Cooking time 15 m...  
13627  Preparation 30 minutes | Cooking 30 minutes | ...  
13628  Preparation 30 minutes | Cooking 30 minutes | ...  
13629  Preparation 3 hours | Cooking 21 minutes | Ser...  
13630  Preparation 1 hour 15 minutes | Cooking 30 min...  
13631  Preparation 5 minutes | Cooking 45 minutes | \...  
13632  Preparation 15 minutes | Cooking 30 minutes | ...  
13633  Preparation 5 minutes | Serves 2\r\n?Add natur...  
13634  Preparation 5 minutes | Cooking 1 hour 40 minu...

In [64]:
#check for missing values
print(combined_df.isnull().sum().sort_values(ascending=False))


Title           0
Ingredients     0
Instructions    0
dtype: int64


- Clean the columns of the combined_df

In [65]:
#write a function to clean the columns
"""
This function should:
1. Clean the Title Column: Remove newlines and extra spaces
2. Clean the Ingredients Column: Convert string ingredients to lists: If they are not already in a list format, convert them and clean any extra spaces.
                                Ensure there are no empty strings or duplicates within each list of ingredients.
3. Clean the Instructions Column: Similar to the Title, ensure that the instructions are clean and properly formatted.
"""

def clean_combined_df(df):
    #title column
    # Remove newlines and extra spaces
    df['Title'] = df['Title'].str.replace('\n', '').str.strip()
    #Ingredients column
    # Remove empty ingredients from each list of ingredients
    df['Ingredients'] = df['Ingredients'].apply(lambda x: [ingredient.strip() for ingredient in x if ingredient.strip() != ''])

    # If there are any duplicate ingredients in each list, remove them
    df['Ingredients'] = df['Ingredients'].apply(lambda x: list(set(x)))

    # Clean the Instructions column
    df['Instructions'] = df['Instructions'].str.replace('\n', '') \
                                       .str.replace('?', '.') \
                                       .str.replace('|', ',') \
                                       .str.replace('\r', '') \
                                       .str.strip()
    #remove   special characters like \x02 from words like  gradu\x02ally
    df['Instructions'] = df['Instructions'].apply(lambda x: re.sub(r'[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\xff]', '', x))#remove special characters
    #if a sentence has more than one fullstop, replace it with one
    df['Instructions'] = df['Instructions'].apply(lambda x: re.sub(r'\.{2,}', '.', x))
    return df

# Use the function on your DataFrame
combined_cleaned = clean_combined_df(combined_df)
combined_cleaned

Title  \
0      Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                        Crispy Salt and Pepper Potatoes   
2                            Thanksgiving Mac and Cheese   
3                     Italian Sausage and Bread Stuffing   
4                                           Newton's Law   
...                                                  ...   
13630                     Bhature\r (Fried Indian Bread)   
13631  Vimumunya vya \rSukari\r (Sweetened Pumpkin & ...   
13632                           Siro\r (Semolina & Nuts)   
13633                           Chaas\r(Diluted Yoghurt)   
13634                                    Groundnut Sauce   

                                             Ingredients  \
0      ['freshly ground pepper'], '2¾ tsp. kosher sal...   
1      ['¾ teaspoon finely ground black pepper', '1 t...   
2      ['1 tsp. garlic powder', '1 cup whole milk', '...   
3      ['2 tablespoons olive oil, '5 garlic cloves, c...   
4      ['Garnish: orange twist and freshly grated or ...   
...                                                  ...   
13630  [salt, coriander, ghee, sugar, fenugreek leave...   
13631    [sugar, pumpkin, coconut milk, cardamon, water]   
13632  [almond, sugar, cow ghee, semolina flour, pist...   
13633                     [salt, water, natural yoghurt]   
13634                [salt, water, sour milk, groundnut]   

                                            Instructions  
0      Pat chicken dry with paper towels, season all ...  
1      Preheat oven to 400F and line a rimmed baking ...  
2      Place a rack in middle of oven; preheat to 400...  
3      Preheat oven to 350F with rack in middle. Gene...  
4      Stir together brown sugar and hot water in a c...  
...                                                  ...  
13630  Preparation 1 hour 15 minutes , Cooking 30 min...  
13631  Preparation 5 minutes , Cooking 45 minutes , S...  
13632  Preparation 15 minutes , Cooking 30 minutes , ...  
13633  Preparation 5 minutes , Serves 2.Add natural y...  
13634  Preparation 5 minutes , Cooking 1 hour 40 minu...  

[13635 rows x 3 columns]

In [66]:
combined_cleaned['Instructions'][13634]

'Preparation 5 minutes , Cooking 1 hour 40 minutes , Serves 4.Place a saucepan over fire and let it preheat.Add the groundnuts, salt and 1/2 a cup of water oras desired.Cook until the water evaporates as you stir gradually. When ready, the nuts produce a pop sound.Once they pop, turn down the heat and continue stirring until the groundnuts are dry (about 13minutes).Remove from heat and allow it to cool down.Using a blender, blend the nuts into a paste. Apestle and mortar can be used in the absence of ablender.Put the groundnut paste into a bowl, add sour milkand stir into thick paste. Water or fresh milk can beused in place of the sour milk.Once ready, put another pan on the heat, add thepeanut paste and stir.Stir until it is smooth but not too thick.Serve hot with green leafy vegetables of yourchoice, fish, sweet potatoes, green bananas, ugali,etc'

### Function to get Recipes based on Ingredients

- Let's see if we can tokenize our Ingredients and see if the system recommends recipes based on the ingredients

In [67]:

def clean_ingredients(ingredients_list):
    # Remove any extra single quotes and fix formatting for each ingredient
    cleaned_list = [re.sub(r"['\"]", "", ingredient) for ingredient in ingredients_list]  # Remove quotes
    cleaned_list = [re.sub(r'\s+', ' ', ingredient) for ingredient in cleaned_list]  # Normalize spaces
    return cleaned_list

# Apply the cleaning function
combined_df['Ingredients'] = combined_df['Ingredients'].apply(clean_ingredients)


# Function to tokenize and normalize ingredients
def tokenize_and_normalize(ingredients_list):
    tokens = []
    for ingredient in ingredients_list:
        # Split ingredient string by commas and strip whitespace
        split_ingredients = [i.strip().lower() for i in ingredient.split(',')]
        
        # Further clean each token: remove unwanted characters
        split_ingredients = [re.sub(r'[^\w\s]', '', i) for i in split_ingredients]  # Remove punctuation
        split_ingredients = [re.sub(r'\s+', ' ', i) for i in split_ingredients]  # Normalize whitespace
        
        # Extend the tokens list with cleaned ingredients
        tokens.extend(split_ingredients)
    
    return tokens


# Apply the function to the Ingredients column
combined_df['Ingredients'] = combined_df['Ingredients'].apply(tokenize_and_normalize)


In [68]:
def recommend_recipes(input_ingredients, combined_df):
    # Normalize user input by stripping whitespace and converting to lower case
    input_ingredients = [ingredient.strip().lower() for ingredient in input_ingredients.split(',')]
    
    # Find matching recipes
    matched_recipes = combined_df[combined_df['Ingredients'].apply(lambda x: any(ingredient in x for ingredient in input_ingredients))]
    
    # Check if any recipes were found
    if matched_recipes.empty:
        return "No recipe found. Try again."
    
    return matched_recipes[['Title','Ingredients', 'Instructions']]


- When the user inputs an ingredient:

In [69]:
# Example user input ingredients
user_input = "thyme"

# Get recommendations
recommended_recipes = recommend_recipes(user_input, combined_df)

# Display recommendations
recommended_recipes


Title  \
273               Braised Celery With Lentils and Garlic   
1479                                   Kale or Chard Pie   
1586                                    Make-Ahead Gravy   
1897      3-Ingredient Garlic-Herb Grilled Chicken Wings   
2766                         Herby Ricotta-Poblano Tacos   
5510                                        Grilled Fish   
5988                                            Gazpacho   
6253                    Duck Breast & Zucchini Tournedos   
6851                                 Chicken & Rice Soup   
7203        Grilled Chicken with Almond and Garlic Sauce   
7808   Rib-Eye, T-Bone, and Strip Steaks Cut Over One...   
8057                                            Gazpacho   
8451                    Heirloom Tomato Herb Pasta Salad   
11968  Sliced Filet Mignon with Fava Beans, Radishes,...   
12036                                Fresh Herb Spaetzle   

                                             Ingredients  \
273    [a small handful hardy herb sprigs such as ros...   
1479   [sliced, rinsed well and thinly sliced, salt a...   
1586   [freshly ground pepper, 6 cups or more thanksg...   
1897   [finely grated or pressed, andor oregano, 2 ta...   
2766   [about 18 teaspoon black pepper, marjoram, 4 g...   
5510   [crumbled, or 12 teaspoon dried thyme or orega...   
5988   [2 tablespoons tomato paste, peeled, 1 medium ...   
6253   [salt, 1 34 oz dried cep mushrooms champagne c...   
6851   [lowsodium chicken broth 4 cups 32 fl oz1 l, c...   
7203   [1 teaspoon coldpressed flaxseed oil, 3 tables...   
7808   [canola oil or vegetable oil, 14 cup finely ch...   
8057   [2 tbsp tomato paste, 3 cups finely diced plum...   
8451   [parsley, thyme, 1 tablespoon whitewine vinega...   
11968  [10 medium radishes, 1 tablespoon canola oil, ...   
12036  [18 teaspoon ground nutmeg, 34 cup whole milk,...   

                                            Instructions  
273    Heat  cup oil in a large high-sided skillet ov...  
1479   Heat the oven to 375F. Melt the butter in a la...  
1586   Heat oil in a large saucepan over medium-high....  
1897   Pat wings very dry with paper towels. Whisk ga...  
2766   On an ungreased griddle or small, heavy skille...  
5510   Combine all the ingredients except fish and le...  
5988   1. In a bowl, reserve 2 tablespoons each of th...  
6253   For the mushroom powder, process the dried cep...  
6851   Combine the broth, rice, and 2 cups (16 fl oz/...  
7203   1. For the sauce, soak the beans in a generous...  
7808   Generously season the steaks with salt and pep...  
8057   1. Reserve 2 tablespoons each of the tomatoes,...  
8451   Whisk together oil, crme frache, vinegar, suga...  
11968  Whisk vinegar and mustard in small bowl. Gradu...  
12036  Blend flour, salt, pepper, and nutmeg in large...

- The results are titles, ingredients and instructions all with the name thyme in it.

- When the user enters a title:

In [70]:
# Example user input ingredients
user_input = "Gazpacho"

# Get recommendations
recommended_recipes = recommend_recipes(user_input, combined_df)

# Display recommendations
recommended_recipes


'No recipe found. Try again.'

In [71]:
#check if there is a recipe with the title 'Gazpacho'
combined_df[combined_df['Title'].str.contains('Gazpacho', case=False)]

Title  \
133                               Speedy Summer Gazpacho   
1771                              Sippin’ Green Gazpacho   
3894                                  Gochujang Gazpacho   
4089                                        Wee Gazpacho   
4476                 Watermelon Gazpacho with Feta Crema   
5297                               Creamy Green Gazpacho   
5988                                            Gazpacho   
6163                     Mixed Berry Gazpacho with Basil   
6260                  Stone Fruit Gazpacho with Scallops   
6564         Fiery Grilled Shrimp with Honeydew Gazpacho   
6765                                            Gazpacho   
7585   Grilled Heirloom Tomato and Mozzarella Sandwic...   
8057                                            Gazpacho   
8230                                Green Gazpacho Shots   
8449                       Tomato and Tomatillo Gazpacho   
8509          Seared Mahi-Mahi with Green Gazpacho Sauce   
8754                            Tomato Cucumber Gazpacho   
8856             Cucumber Gazpacho with Shrimp and Melon   
9405                                  Farmstand Gazpacho   
9415                                      Gazpacho Salsa   
9805                           Tomatillo-Pepita Gazpacho   
10722                Grilled Tomato-Bell Pepper Gazpacho   
10765       Ziti with Grilled-Gazpacho Sauce and Sausage   
11254                              Gazpacho Bloody Marys   
11674                              Celery Apple Gazpacho   
11699                 Grilled Gazpacho Salad with Shrimp   
12499                                    Speedy Gazpacho   

                                             Ingredients  \
133    [2 garlic cloves, 1 cucumber, grated, plus ext...   
1771   [2 garlic cloves, 34 cup or more olive oil, pa...   
3894   [2 tablespoons chopped fresh parsley, 2 garlic...   
4089   [1 teaspoon salt, peeled, 1 teaspoon lemon jui...   
4476   [peeled, 1 large beefsteak tomato, kosher salt...   
5297   [kosher or sea salt, 3 large basil leaves, 1 s...   
5988   [2 tablespoons tomato paste, peeled, 1 medium ...   
6163   [cut into 12 pieces, extravirgin olive oil for...   
6260   [pitted, coarsely chopped, 1 shallot, espelett...   
6564   [seeded, extravirgin olive oil, 3 tablespoons ...   
6765   [1 14 oz white countrystyle bread without crus...   
7585   [1 teaspoon or more coarse kosher salt, 8 13in...   
8057   [2 tbsp tomato paste, 3 cups finely diced plum...   
8230   [quartered, peeled, 1 medium green bell pepper...   
8449   [1 12 pounds tomatoes, quartered, and quartere...   
8509   [5 12 tablespoons olive oil, divided, 34 cup c...   
8754   [2 tablespoons finely chopped fresh basil plus...   
8856   [34 teaspoon kosher salt, peeled, and cut into...   
9405   [2 cups tomato juice, 13 cup extravirgin olive...   
9415   [3 medium tomatoes, peeled, 1 small cucumber, ...   
9805   [1 large fresh poblano chile, 2 pounds tomatil...   
10722  [peeled, 34 cup about cold water optional, at ...   
10765  [1 12 pounds cherry tomatoes, 1 pound ziti, di...   
11254  [2 28ounce cans italian tomatoes such as san m...   
11674  [1 granny smith apple, 1 12 cups cold water, 2...   
11699  [peeled, 1 yellow bell pepper, 1 teaspoon hot ...   
12499  [seeded, peeled, 1 english hothouse cucumber a...   

                                            Instructions  
133    Put all the ingredients in a food processor an...  
1771   Pure cucumbers, garlic, and 1/2 cup water in a...  
3894   Pulse 1 1/2 cups tomato, 1/4 cup cucumber, ora...  
4089   Place all the ingredients in a food processor ...  
4476   Pure watermelon, tomato, cucumber, jalapeo, oi...  
5297   Reserve one-quarter of the tomato, two cucumbe...  
5988   1. In a bowl, reserve 2 tablespoons each of th...  
6163   Combine first 6 ingredients and basil sprig in...  
6260   Combine the peaches, plums, watermelon, garlic...  
6564   Cut away the honeydew rind and the dark green ...  
6765   Peel and cut each garlic clove in

- They get the response "No recipe found.Try again" but if you check among the foods, there is a recipe for Gazpacho.

- In the above function we see that when a user enters an ingredient it results in sample recipes the user can cook. However when the user tries to enter a title of a food they want, it returns the response "No recipe found. Try Again."
- If you search the title of the food in the list of food, it's there. So the function isn't really working accordingly.
- Thus we can try Feature Engineering. Create a new column with combined title and ingredients, Vectorize, create cosine similarities and see how well it performs compared to our function above.

### Feature Engineering

- Create the column containing combined title and ingredient.

In [72]:
# Combining title and ingredients
"""
This code is meant to combine both the ingredients and the title into a 
single string for each recipe this will allow to system to search both 
the ingredients and the title since ther user can input any of them.
"""
combined_df['Title_Ingredients'] = combined_df['Title'] + ' ' + combined_df['Ingredients'].apply(lambda x: ' '.join(x))
combined_df.head()

Title  \
0  Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                    Crispy Salt and Pepper Potatoes   
2                        Thanksgiving Mac and Cheese   
3                 Italian Sausage and Bread Stuffing   
4                                       Newton's Law   

                                         Ingredients  \
0  [freshly ground pepper, 2¾ tsp kosher salt, 1 ...   
1  [¾ teaspoon finely ground black pepper, 1 teas...   
2  [1 tsp garlic powder, 1 cup whole milk, 1 tsp ...   
3  [2 tablespoons olive oil, 5 garlic cloves, cas...   
4  [garnish orange twist and freshly grated or gr...   

                                        Instructions  \
0  Pat chicken dry with paper towels, season all ...   
1  Preheat oven to 400F and line a rimmed baking ...   
2  Place a rack in middle of oven; preheat to 400...   
3  Preheat oven to 350F with rack in middle. Gene...   
4  Stir together brown sugar and hot water in a c...   

                                   Title_Ingredients  
0  Miso-Butter Roast Chicken With Acorn Squash Pa...  
1  Crispy Salt and Pepper Potatoes ¾ teaspoon fin...  
2  Thanksgiving Mac and Cheese 1 tsp garlic powde...  
3  Italian Sausage and Bread Stuffing 2 tablespoo...  
4  Newton's Law garnish orange twist and freshly ...

In [73]:
# preview combined column of title together with ingredients
combined_df['Title_Ingredients'][0]

'Miso-Butter Roast Chicken With Acorn Squash Panzanella freshly ground pepper 2¾ tsp kosher salt 1 tbsp finely chopped rosemary melted 6 tbsp unsalted butter 3 tbsp apple cider vinegar ½ small red onion 2 tsp white miso divided 2 tbsp finely chopped sage plus more 1 3½4lb whole chicken ¼ cup allpurpose flour plus 3 tbsp room temperature kosher salt 2 small acorn squash about 3 lb total 1 tbsp white miso 2 medium apples such as gala or pink lady about 14 oz total torn into 1 pieces about 2½ cups ¼ tsp ground allspice cut into 1 pieces ⅓ loaf goodquality sturdy white bread 2 tbsp extravirgin olive oil pinch of crushed red pepper flakes room temperature thinly sliced ¼ cup dry white wine cored freshly ground black pepper 2 cups unsalted chicken broth 2 tbsp unsalted butter'

### TF-IDF Vectorizer
-The TF-IDF vectorizer will then create vectors based on both the recipe title and ingredients. This will give you a more comprehensive representation of the recipe.

In [74]:
# intiliazing TFIDF Vectorizer
tfidf_vectorizer = TfidfVectorizer()
# fit the vectorizer
tfidf_matrix = tfidf_vectorizer.fit_transform(combined_df['Title_Ingredients'])


- Check the shape of the matrix and see the features to know how the vectorization has taken place.

In [75]:
# Check the shape of the TF-IDF matrix (rows x features)
print(f'Shape of matrix:{tfidf_matrix.shape}')

# Get the feature names (terms)
tfidf_feature_names = tfidf_vectorizer.get_feature_names_out()
print(f'Feature Names:{tfidf_feature_names[:10]}')  # Show first 10 features (terms)

Shape of matrix:(13635, 11974)
Feature Names:['00' '000mg' '011' '018' '018oz' '02' '025' '028' '035' '037']


- In the above we can see how the shape of the matrix is 13,635 by 13,690 after vectorization and sample feature names.

- Let's get the cosine Similarities

In [76]:
#cosine similarity
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

#print the cosine similarity matrix shape
print(f'Cosine Similarity Matrix Shape: {cosine_sim.shape}')


Cosine Similarity Matrix Shape: (13635, 13635)


- Let's see how well our system recommends recipes based on the cosine similarities.
- Let's take a look at the first recipe and the top 5 recipes with highest cosine similarities with it.

In [77]:
#example getting top 5 most similar recipes with the first recipe
similar_recipes = cosine_sim[0].argsort()[-6:-1] #


In [78]:
#display the first recipe
combined_df.iloc[0][['Title', 'Ingredients', 'Instructions']]

Title           Miso-Butter Roast Chicken With Acorn Squash Pa...
Ingredients     [freshly ground pepper, 2¾ tsp kosher salt, 1 ...
Instructions    Pat chicken dry with paper towels, season all ...
Name: 0, dtype: object

In [79]:
# Display the similar recipes
combined_df.iloc[similar_recipes][['Title', 'Ingredients', 'Instructions']][::-1]

Title  \
590   Grilled Whole Cauliflower with Miso Mayo   
129      Grilled Potatoes With Red Miso Butter   
331           Risotto With Mushrooms and Thyme   
4237                     Bacon-Cheddar Muffins   
26                           Miso-Squash Ramen   

                                            Ingredients  \
590   [1 tbsp soy sauce, 1 tbsp ketchup, 4 tbsp unsa...   
129   [finely chopped, 4 tbsp unsalted butter, 1 gar...   
331   [2 cups carnaroli or japanese sushi rice, trim...   
4237  [1 cup plus 2 tbsp buttermilk, 2 tbsp sugar, 6...   
26    [handful of cilantro leaves with tender stems,...   

                                           Instructions  
590   Prepare a grill for medium-high heat. Sprinkle...  
129   Prepare a grill for medium-high heat. Bring po...  
331   Combine 1 Tbsp. salt and 10 cups water in a me...  
4237  Preheat oven to 400 with rack near top. Line t...  
26    Preheat oven to 400F. Whisk oil, miso, maple s...

In [80]:
#display the similarities of the first recipe with the top 5 similar recipes
cosine_sim[0][similar_recipes][::-1].round(4)

array([0.4927, 0.4799, 0.4646, 0.4638, 0.4568])

- In the above, there have been 5 top recipe recommendations  based on their cosine similarities with the first recipe. The first recipe having the highest similarity of 0.4927

- Clean the title.

In [81]:
# clean title column
def clean_input(title):
    # Normalize titles by removing punctuation, converting to lowercase, and stripping extra spaces
    title = re.sub(r'[^\w\s]', '', title)
    title = re.sub(r'\s+', ' ', title).strip().lower()
    return title
combined_df['Title'] = combined_df['Title'].apply(clean_input)

- Create a function to recommend the recipes.

In [82]:
#function to recommend recipes
def recommend_recipe(user_input, combined_df, tfidf_vectorizer, tfidf_matrix, n=10):
    # Clean the user input
    user_input_cleaned = clean_input(user_input)
    
    # Check if the user input matches a specific title
    matching_title = combined_df[combined_df['Title'].str.lower().str.contains(user_input_cleaned)]
    
    #check if there is a match by title
    if not matching_title.empty:
        return matching_title[['Title', 'Ingredients', 'Instructions']]
    
    # If no exact title match is found, proceed with TF-IDF based recommendation
    # Transform the user input
    user_input_transformed = tfidf_vectorizer.transform([user_input_cleaned])
    
    # Compute the cosine similarity between the user input and the recipes
    cosine_sim = cosine_similarity(user_input_transformed, tfidf_matrix)
    
    # Get the index of the top n most similar recipes
    similar_recipes = cosine_sim[0].argsort()[-n:][::-1]
    
    # Return the top n most similar recipes
    return combined_df.iloc[similar_recipes][['Title', 'Ingredients', 'Instructions']]
    

In [83]:
#Example usage
user_input = "eggs,tomatoes,onions"
recommendations = recommend_recipe(user_input, combined_df, tfidf_vectorizer, tfidf_matrix)

# View recommendations
recommendations

Title  \
13634                                    groundnut sauce   
4552          charred and raw corn with chile and cheese   
4550            grilled rib eye with paprika vinaigrette   
4549   grilled zucchini and leeks with walnuts and herbs   
4548            asparagus with bacon and hardboiled eggs   
4547                     mint and cuminspiced lamb chops   
4546                         moroccan chicken brochettes   
4545                                grilled sesame squid   
4544              plum tarts with honey and black pepper   
4543                       pickled carrots with tarragon   

                                             Ingredients  \
13634                [salt, water, sour milk, groundnut]   
4552   [husked, 2 ounces fresh cotija cheese or queso...   
4550   [finely chopped, 2 tablespoons light brown sug...   
4549   [1 garlic clove, divided, white and palegreen ...   
4548   [1 tablespoon pure maple syrup, finely chopped...   
4547   [quartered, 24 untrimmed lamb rib chops about ...   
4546   [boneless chicken thighs, and fresh mint leave...   
4545   [2 tablespoons toasted sesame oil, 1 12 pounds...   
4544   [flaky sea salt such as maldon, 1 sheet frozen...   
4543   [quartered, 2 red or green thai chiles, peeled...   

                                            Instructions  
13634  Preparation 5 minutes , Cooking 1 hour 40 minu...  
4552   Prepare grill for medium heat. Cut kernels fro...  
4550   Heat oil and paprika in a small saucepan over ...  
4549   Prepare grill for medium-high heat. Toast waln...  
4548   Cook bacon in a large skillet over medium heat...  
4547   Pulse onions, cilantro, parsley, mint, cumin, ...  
4546   Place garlic in a mortar; season with salt and...  
4545   Prepare grill for high heat; oil grate with ve...  
4544   Preheat oven to 425. Cut pastry into six 4" sq...  
4543   Pack carrots, shallot, garlic, chiles, tarrago...

- Based on the above, the system works in a way that when a user enters a specific title of a food like "mandazi" it returns all instances where the food title is seen otherwise, if the user enters ingredients, it converts the user input to a vector then performs the cosine similarity. Afterwards, it returns recipes of the top foods with highest cosine similarities.

- This system works accordingly in recommending the recipes to cook.


### Saving Code
- In order for a user to get recommended recipes, we need to save some code such as the recommend recipe function, combined_df, the tf_idf vectorizer as
.py and pickle files in order for a user to get feedback whenever they input data.
- So the TF-IDF Vectorizer and the combined dataframe will be saved as pickle files while the combined_df saved as .py file to make it easy to adjust here and there.

In [84]:
#saving the model
import joblib

#save the tfidf vectorizer
joblib.dump(tfidf_vectorizer, 'tfidf_vectorizer.pkl')

#save the tfidf matrix
joblib.dump(tfidf_matrix, 'tfidf_matrix.pkl')

#save the cosine similarity matrix
joblib.dump(cosine_sim, 'cosine_sim.pkl')

#save the combined dataframe
combined_df.to_csv('combined_df.csv', index=False)
